test/common/router/config_impl_test.cc

```diff
@@ -4204,6 +4209,148 @@ name: foo
               config.route(headers, 0)->routeEntry()->clusterName());
   }
 }
+
+class PerFilterConfigsTest : public testing::Test,
+                             public Extensions::HttpFilters::Common::EmptyHttpFilterConfig {
+public:
+  void SetUp() override {
+    registered_factory_.reset(
+        new Registry::InjectFactory<Server::Configuration::NamedHttpFilterConfigFactory>(*this));
+  }
+
+  // EmptyHttpFilterConfig
+  Server::Configuration::HttpFilterFactoryCb
+  createFilter(const std::string&, Server::Configuration::FactoryContext&) override {
+    NOT_IMPLEMENTED;
+  }
+  ProtobufTypes::MessagePtr createEmptyRouteConfigProto() override {
+    return ProtobufTypes::MessagePtr{new ProtobufWkt::Timestamp()};
+  }
+  std::string name() override { return "test.filter"; }
+
+  void checkEach(const std::string& yaml, uint32_t expected_entry, uint32_t expected_route,
+                 uint32_t expected_vhost) {
+    const ConfigImpl config(parseRouteConfigurationFromV2Yaml(yaml), runtime_, cm_, true);
+
+    const auto route = config.route(genHeaders("www.foo.com", "/", "GET"), 0);
+    const auto* route_entry = route->routeEntry();
+    const auto& vhost = route_entry->virtualHost();
+
+    check(route_entry->perFilterConfig(name()), expected_entry, "route entry");
+    check(route->perFilterConfig(name()), expected_route, "route");
+    check(vhost.perFilterConfig(name()), expected_vhost, "virtual host");
+  }
+
+  void check(const Protobuf::Message* cfg, uint32_t expected_seconds, std::string source) {
+    EXPECT_NE(nullptr, cfg) << "config should not be null for source: " << source;
+    EXPECT_NO_THROW({
+      const auto ts = dynamic_cast<const ProtobufWkt::Timestamp*>(cfg);
+      EXPECT_EQ(expected_seconds, ts->seconds())
+          << "config value does not match expected for source: " << source;
+    }) << "config should properly dynamic_cast to the appropriate type for source: "
+       << source;
+  }
+
+  std::unique_ptr<Registry::InjectFactory<Server::Configuration::NamedHttpFilterConfigFactory>>
```

Comment:
nit: I think this can just be stack allocated with the test? Do you need a unique_ptr?